program/src/processor.rs

```diff
@@ -5313,6 +5318,286 @@ impl Processor {
+
+        let cpi_ix = Instruction {
+            program_id: *cpi_prog_ai.key,
+            data: cpi_data,
+            accounts: cpi_account_metas,
+        };
+
+        let signers_seeds = gen_signer_seeds(&mango_group.signer_nonce, mango_group_ai.key);
+        solana_program::program::invoke_signed(&cpi_ix, &cpi_account_infos, &[&signers_seeds])?;
+
+        // Determine change in vaults and apply to MangoAccount
+        for (i, bank_set) in bank_sets.iter_mut().enumerate() {
+            let pre_amount = I80F48::from_num(pre_amounts[i]);
+            let token_account = TokenAccount::load(bank_set.vault_ai)?;
+            token_account.check_mango_reqs(&mango_group)?; // make sure only amount changed
+            let post_amount = I80F48::from_num(token_account.0.amount);
+            // let post_amount =
+            //     I80F48::from_num(Account::unpack(&bank_set.vault_ai.try_borrow_data()?)?.amount);
+            let change = post_amount - pre_amount;
+            checked_change_net(
+                &mango_cache.root_bank_cache[bank_set.token_index],
+                &mut bank_set.node_bank,
+                &mut mango_account,
+                mango_account_ai.key,
+                bank_set.token_index,
+                change,
+            )?;
+            health_cache.update_token_val_orders_vec(
+                &mango_group,
+                &mango_cache,
+                &mango_account,
+                &open_orders_ais,
+                bank_set.token_index,
+            )?;
+            // TODO log
+        }
+        let post_health = health_cache.get_health(&mango_group, HealthType::Init);
+        // If an account is in reduce_only mode, health must only go up
+        check!(
+            post_health >= ZERO_I80F48 || (reduce_only && post_health >= pre_health),
+            MangoErrorCode::InsufficientFunds
+        )?;
+
+        Ok(())
+    }
+
+    fn flash_loan(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        liquidity_amount: u64,
+        cpi_data: Vec<u8>,
+    ) -> MangoResult {
+        const NUM_FIXED: usize = 6;
+        let (fixed_ais, remaining_ais) = array_refs![accounts, NUM_FIXED; ..;];
+        let [
+            mango_group_ai,           // read
+            signer_ai,                // read
+            source_token_ai,          // write
+            destination_token_ai,     // write
+            token_prog_ai,            // read
+            flash_loan_prog_ai,       // read
+        ] = fixed_ais;
+
+        check!(program_id != flash_loan_prog_ai.key, MangoErrorCode::InvalidProgramId)?;
+
+        let mango_group = MangoGroup::load_checked(mango_group_ai, program_id)?;
+        let source_token = Account::unpack(&source_token_ai.try_borrow_data()?)?;
+
+        let flash_loan_amount = if liquidity_amount == u64::MAX {
+            source_token.amount
+        } else {
+            min(source_token.amount, liquidity_amount)
```

Comment:
as written, it could just always be `min()` without special casing for `u64::MAX`. But probably cases where liquidity_amount < source_token.amount should error?